src/hotspot/share/runtime/interfaceSupport.inline.hpp

```diff
@@ -236,12 +213,16 @@ class ThreadBlockInVMPreprocess : public ThreadStateTransition {
       SafepointMechanism::process_if_requested(_thread, _allow_suspend);
     }
   }
+};
 
+class ThreadBlockInVM  : public ThreadBlockInVMPreprocess<> {
+ public:
+  ThreadBlockInVM(JavaThread* thread, bool allow_suspend = false)
+    : ThreadBlockInVMPreprocess(thread, emptyOp, allow_suspend) {}
+ private:
   static void emptyOp(JavaThread* current) {}
 };
 
-typedef ThreadBlockInVMPreprocess<> ThreadBlockInVM;
-
```

Comment:
Now we can use ThreadBlockInVM in wait_for_object_deoptimization(), but unlike the other uses, in here we need to allow suspends since we could be going back to Java. With the current definition of ThreadBlockInVM we would be forced to pass something for the PRE_PROC parameter which doesn't make sense. Rearranging the parameters for the ThreadBlockInVMPreprocess constructor would mess up the other callers, and creating a new constructor with default parameter allow_suspend and no PRE_PROC would be too confusing I think. So I think this is the cleanest solution. ThreadBlockInVM is just a type of ThreadBlockInVMPreprocess with no callback.

Reply:
> Now we can use ThreadBlockInVM in wait_for_object_deoptimization(), but unlike
> the other uses, in here we need to allow suspends since we could be going back
> to Java.

Thanks for the explanation. I almost made the mistake of thinking it was ok not
to allow suspends in wait_for_object_deoptimization() but not allowing suspends
on the suspendee side of course does not mean a suspender cannot complete a
suspend request while the suspendee thread is blocked for object
deoptimization. This is obvious to me now but somehow I was really confused by
the name "allow_suspend".